### PR TITLE
Efficient /state_ids

### PR DESCRIPTION
**Pull request by neilalexander**
Beforehand, `/state_ids` was still pulling all of the state and auth events out of the database and unmarshalling them. This isn't really necessary because the `roomserver_events` table already knows both the event ID and the auth event NIDs, so we can look those up instead. This adds the necessary roomserver query shape to do just that.

### CODE REVIEW

File: roomserver/api/query.go [hunk 1]
```diff
@@ -250,6 +250,28 @@ type QueryStateAndAuthChainResponse struct {
...
+type QueryStateAndAuthChainIDsResponse struct {
+	// Does the room exist on this roomserver?
+	// If the room doesn't exist this will be false and StateEvents will be empty.
+	RoomExists bool `json:"room_exists"`
+	// The room version of the room.
+	RoomVersion gomatrixserverlib.RoomVersion `json:"room_version"`
+	// The state and auth chain event IDs that were requested.
+	// The lists will be in an arbitrary order.
+	StateEvents     []string `json:"state_event_ids"`
+	AuthChainEvents []string `json:"auth_chain_event_ids"`
```
Comment (kegsay) [inline roomserver/api/query.go:272]: These aren't events, they are event IDs. `AuthChainEventIDs` and `StateEventIDs` please.

File: roomserver/internal/query/query.go [hunk 2]
```diff
@@ -526,6 +526,70 @@ func (r *Queryer) QueryStateAndAuthChain(
...
+	if err != nil {
+		return fmt.Errorf("r.DB.EventIDs: %w", err)
+	}
+
+	for _, eventID := range stateEventIDs {
+		response.StateEvents = append(response.StateEvents, eventID)
+	}
+
+	for _, eventID := range authEventIDs {
+		response.AuthChainEvents = append(response.AuthChainEvents, eventID)
```
Comment (kegsay) [inline roomserver/internal/query/query.go:587]: These are just the auth events for the state events, not the auth chain.

File: roomserver/storage/sqlite3/events_table.go [hunk 3]
```diff
@@ -145,6 +149,7 @@ func prepareEventsTable(db *sql.DB) (tables.Events, error) {
 		{&s.bulkSelectEventIDStmt, bulkSelectEventIDSQL},
 		{&s.bulkSelectEventNIDStmt, bulkSelectEventNIDSQL},
 		//{&s.selectRoomNIDForEventNIDStmt, selectRoomNIDForEventNIDSQL},
+		//{&s.bulkSelectEventAuthEventNIDsStmt, bulkSelectEventAuthEventNIDsSQL},
```
Comment (kegsay) [inline roomserver/storage/sqlite3/events_table.go:152]: Spurious, please remove

**Review by kegsay — CHANGES_REQUESTED** [review]
It isn't returning the auth chain.

**Comment by Devkey19** [comment]
Bug

**Comment by neilalexander** [comment]
Stale.

**Pull request closed without merging**